src/charm.py

```diff
@@ -738,10 +845,105 @@ def _get_initialization_secret_from_peer_relation(self) -> Tuple[str, List[str]]
+                aws_access_key_id=s3_parameters["access-key"],
+                aws_secret_access_key=s3_parameters["secret-key"],
+                region_name=s3_parameters["region"],
+            )
+            session.resource("s3", endpoint_url=s3_parameters["endpoint"])
+            return session
+        except KeyError as e:
+            logger.warning("Missing required S3 parameter: %s", e)
+            return None
+        except ValueError as e:
+            logger.warning("Error creating resource: %s", e)
+            return None
+
+    def _get_missing_s3_parameters(self) -> List[str]:
+        """Returns list of missing S3 parameters.
+
+        Returns:
+            List[str]: List of missing required S3 parameters.
+        """
+        s3_parameters = self.s3.get_s3_connection_info()
+        return [param for param in REQUIRED_S3_PARAMETERS if param not in s3_parameters]
+
+    def _retrieve_s3_parameters(self) -> Dict[str, str]:
+        """Retrieve S3 parameters from the S3 integrator relation.
+
+        Removes leading and trailing whitespaces from the parameters.
+        Adds default region if not provided.
+
+        Returns:
+            Dict[str, str]: Dictionary of the S3 parameters.
+        """
+        s3_parameters = self.s3.get_s3_connection_info()
+        if "region" not in s3_parameters:
+            s3_parameters["region"] = DEFAULT_REGION
+        for key, value in s3_parameters.items():
+            if isinstance(value, str):
+                s3_parameters[key] = value.strip()
+
+        return s3_parameters
+
+    def _create_s3_bucket(
+        self,
+        session: boto3.session.Session,
+        bucket_name: str,
+        endpoint: str,
+        region: str,
+    ) -> None:
+        """Create S3 bucket.
+
+        If the bucket already exists, it will be skipped.
+
+        Args:
+            session: S3 session.
+            region: S3 region.
+            bucket_name: S3 bucket name.
+            endpoint: S3 endpoint.
+        """
+        s3 = session.resource("s3", endpoint_url=endpoint)
+        try:
+            # Checking if bucket already exists
+            bucket = s3.Bucket(bucket_name)
+            bucket.meta.client.head_bucket(Bucket=bucket_name)
+            logger.info("Bucket %s exists.", bucket_name)
+            return
+        except ClientError:
+            logger.info("Bucket %s doesn't exist, creating it.", bucket_name)
+            pass
+        except BotoCoreError as e:
+            logger.warning("Failed to check wether bucket exists. %s", e)
+            raise e
+        try:
+            if region == DEFAULT_REGION:
+                bucket = bucket.create()
+            else:
+                bucket.create(CreateBucketConfiguration={"LocationConstraint": region})
+            bucket.wait_until_exists()
+        except (BotoCoreError, ClientError) as error:
+            logger.warning("Couldn't create bucket named '%s' in region=%s.", bucket_name, region)
```

Comment:
I think we should log at the `error` level here, we are failing the action right after.

Reply:
✅